### PR TITLE
Special handling for 1D covolutional kernels in cuDNN flavor of conv_op.

### PR DESCRIPTION
**Pull request by drnikolaev**
Essentially makes cuDNN to think of those kernels like of Nx1 ones.

### CODE REVIEW

**Comment by ezyang** [comment]
Can you elaborate more about this change? Is it a performance improvement? (If so, can you give some before and after numbers?) Is it a correctness fix? (If so, can you add a test?)

Thanks!

**Comment by drnikolaev** [comment]
Sure! It's definitely not perf improvement. It is correctness fix. The problem is that some cuDNN calls (like `cudnnGetConvolutionForwardAlgorithm` for example) assert the kernel to have either 2 or 3 dimensions. So, PyTorch crashes otherwise. Not good when user wants 1D kernel. The fix doesn't change the state, it just re-declares 1D kernels of size N to be considered as 2D kernels of size Nx1. Let me compose a test for this use case, please hold on...

**Comment by drnikolaev** [comment]
@ezyang it's actually already there: `caffe2/python/operator_test/conv_test.py::TestConvolution::test_convolution_layout` crashes before the fix and succeeds after it. But seems like caffe2 tests never get executed. Well, at least I couldn't find any trace of them (check here for example: https://circleci.com/gh/pytorch/pytorch/92474?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link ). Am I missing something? If those tests are still ignored, may be it's time to add them little by little? In this case what driver would be better to use, `test/run_test.sh` or `.jenkins/caffe2/test.sh`?

**Comment by ezyang** [comment]
The merge was easy, I fixed it for you. Thank you!

**Comment by ezyang** [comment]
Tests are failing on CI.

**Comment by drnikolaev** [comment]
Trying to reproduce the failure, please hold on..

**Comment by drnikolaev** [comment]
@ezyang That failures are reproduced and fixed now (sorry, can't squash my commits any longer). They are about forcing algos which are not supported for given parameters (like I mentioned in my previous comment). Now Caffe2 handles such use cases properly: if algorithm chosen by user is not supported, it falls back to default one (determinism honored) and issues a warning. It's more user friendly than crash (NVCaffe actually does the same).

File: caffe2/operators/conv_op_cudnn.cc [hunk 1]
```diff
@@ -758,14 +758,32 @@ bool CudnnConvOp::DoRunWithType() {
...
     }
-    CUDNN_ENFORCE(cudnnGetConvolutionForwardWorkspaceSize(
-        cudnn_wrapper_.inline_cudnn_handle(),
-        bottom_desc_,
-        filter_desc_,
-        conv_desc_,
-        top_desc_,
-        algo_,
-        &cudnn_ws_nbytes_));
+    for(int step = 0; step < 2; ++step) {
```
[the remark this answers is not in the document]
Comment (BIT-silence) [reply caffe2/operators/conv_op_cudnn.cc]: I see, thanks for the clarification.

**Review by xiaomengy — APPROVED** [review]

**Pull request closed without merging**